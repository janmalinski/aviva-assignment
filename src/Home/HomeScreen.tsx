import React, { useCallback, useEffect, useState } from 'react';
import { StyleSheet, Alert } from 'react-native';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { getData, storeData, removeData } from '../helpers/Storage';

import { getWeatherDetails } from './homeApi';
import { TRootNavigatorParams, TWeatherDetails } from '../navigation/RootNavigator';
import WeatherDetails from '../components/WeatherDetails';
import SearchInput from './components/SearchInput';
import SaveCityName from './components/SaveCityName';
import Button from '../components/Button';
import { debounce } from '../helpers/Debounce';

const HomeScreen = () => {
  const [text, setText] = useState<string>('');
  const [weatherData, setWeatherData] = useState<TWeatherDetails | null>(null);
  const [cityNameSaved, setCityNameSaved] = useState<boolean>(false);

  const navigation = useNavigation<NavigationProp<TRootNavigatorParams, "Home">>();

  useEffect(() => {
      getDataHandler();
  }, []);

  useEffect(() => {
    if(cityNameSaved){
      storeData('@city_name', text);
    } else if(text !== '' && !cityNameSaved){
      removeData('@city_name');
    }
  }, [cityNameSaved]);

  useEffect(() => {
    if(text !== ''){
      debounceGetWeatherData();
    }
    if(!text){
      setCityNameSaved(false);
    }
  }, [text]);

  const getDataHandler = async() => {
    const data = await getData('@city_name');
    if(data) {
      setText(data as string);
      setCityNameSaved(true);
    }
  };

  const getWeatherData = async() => {
    const data = await getWeatherDetails(text);
    setWeatherData(data);
  }

  const debounceGetWeatherData = debounce(getWeatherData, 500)
  
  const handleSetText =  useCallback( async(value: string) => {
    setText(value);
  }, [setText]);

  const navigateToWeatherDetails = useCallback(() => {
    if(weatherData && text){
      navigation.navigate("WeatherDetails", {...weatherData});
      !cityNameSaved && setText('');
    } else {
      Alert.alert("You didn't type full city name.", "Please type city name and then click search button.");
    }
  }, [weatherData, text, navigation, cityNameSaved, setText]);

  const handleSaveCity = useCallback(() => {
    setCityNameSaved(prevState => !prevState); 
  },[setCityNameSaved]);

  return (
      <SafeAreaView style={styles.container} edges={['top']}>
        <SearchInput text={text} handleSetText={handleSetText} navigateToWeatherDetails={navigateToWeatherDetails} />
          <SaveCityName disabled={text === '' ? true : false} cityNameSaved={cityNameSaved} handleSaveCity={handleSaveCity}/>
          {weatherData && text && (
            <>
              <WeatherDetails weatherData={weatherData} screen='Home' />
              <Button title='Get more details' onPress={navigateToWeatherDetails} />
            </>
            )}
      </SafeAreaView>   
  );
};

export default HomeScreen;

const styles = StyleSheet.create({
    container: {
      flex: 1,
      paddingHorizontal: 40,
      backgroundColor: '#e4f6f4'
    },
});