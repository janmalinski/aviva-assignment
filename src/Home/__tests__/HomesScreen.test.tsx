import React from "react";
import { render } from "@testing-library/react-native";
import HomeScreen from "../HomeScreen";

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
    }),
  };
});

describe('HomeScreen', () => {
    it('Should render correctly', () => {
        render(<HomeScreen />)
    });

    it('Should match a snapshot', () => {
      const elJson = render(<HomeScreen />).toJSON();
      expect(elJson).toMatchSnapshot(); 
  });
});



