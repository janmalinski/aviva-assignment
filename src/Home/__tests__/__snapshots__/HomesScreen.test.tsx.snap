// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Should match a snapshot 1`] = `
<RNCSafeAreaView
  edges={
    [
      "top",
    ]
  }
  style={
    {
      "backgroundColor": "#e4f6f4",
      "flex": 1,
      "paddingHorizontal": 40,
    }
  }
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "flexDirection": "row",
          "height": 40,
          "marginBottom": 15,
          "marginTop": 30,
        },
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Search"
      style={
        {
          "borderBottomWidth": 1,
          "flex": 1,
          "height": 40,
          "paddingHorizontal": 5,
        }
      }
      testID="search-text-input"
      value=""
    />
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 30,
          },
          {
            "alignSelf": "center",
          },
          {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="search-button"
    >
      ï”‰
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#000",
          "borderWidth": 1,
          "height": 16,
          "justifyContent": "center",
          "marginRight": 5,
          "width": 16,
        }
      }
    />
    <Text
      style={
        {
          "fontSize": 16,
        }
      }
    >
      Save name of favourite city
    </Text>
  </View>
</RNCSafeAreaView>
`;
